File: 0116-populating-next-right-pointers-in-each-node/0116-populating-next-right-pointers-in-each-node.js
/**
 * // Definition for a _Node.
 * function _Node(val, left, right, next) {
 *    this.val = val === undefined ? null : val;
 *    this.left = left === undefined ? null : left;
 *    this.right = right === undefined ? null : right;
 *    this.next = next === undefined ? null : next;
 * };
 */

/**
 * @param {_Node} root
 * @return {_Node}
 */
var connect = function(root) {
    if (root==null){
        return root
    }
    const q=[]
    q.push(root)
    while(q.length>0){
        let qlength = q.length;
        for(let i=0;i<qlength;i++){
            let temp=q[0];
            q.shift();
            if (i!=qlength-1){
                temp.next = q[0]
            }
            if(temp.left){
                q.push(temp.left)
            }
            if (temp.right){
                q.push(temp.right)
            }
        }
    }
    return root
    
};

File: 0116-populating-next-right-pointers-in-each-node/0116-populating-next-right-pointers-in-each-node.cpp
/*
// Definition for a Node.
class Node {
public:
    int val;
    Node* left;
    Node* right;
    Node* next;

    Node() : val(0), left(NULL), right(NULL), next(NULL) {}

    Node(int _val) : val(_val), left(NULL), right(NULL), next(NULL) {}

    Node(int _val, Node* _left, Node* _right, Node* _next)
        : val(_val), left(_left), right(_right), next(_next) {}
};
*/

class Solution {
public:
    Node* connect(Node* root) {
        if(root==NULL){
            return root;
        }

        queue<Node*> myQueue;
        myQueue.push(root);
        while (myQueue.size() > 0) {
            int myQueueSize = myQueue.size();
            for (int i = 0; i < myQueueSize; i++) {
                Node* temp = myQueue.front();
                myQueue.pop();
                if (i != myQueueSize - 1) {
                    temp->next = myQueue.front();
                }

                if (temp->left != NULL) {
                    myQueue.push(temp->left);
                }
                if (temp->right != NULL) {
                    myQueue.push(temp->right);
                }
            }
        }
        return root;
    }
};